ams/server/src/main/java/com/netease/arctic/server/optimizing/plan/OptimizingPlanner.java

```diff
@@ -60,8 +66,14 @@ public OptimizingPlanner(
     super(tableRuntime, table);
     this.partitionFilter =
         tableRuntime.getPendingInput() == null
-            ? null
-            : tableRuntime.getPendingInput().getPartitions()::contains;
+            ? Expressions.alwaysTrue()
+            : tableRuntime.getPendingInput().getPartitions().entrySet().stream()
+                .map(
+                    entry ->
+                        ExpressionUtil.convertPartitionDataToDataFilter(
+                            table, entry.getKey(), entry.getValue()))
+                .reduce(Expressions::and)
```

Comment:
It should be `Expressions::or` for partitions of different specId.

Reply:
You're right, done.